In [1]:
import pandas as pd
import os
print(os.getcwd())
%matplotlib notebook
%matplotlib inline
import matplotlib.pyplot as plt
plt.style.use('dark_background')
import numpy as np

C:\Users\bhara\MachineLearning


##### Loading the dataset
Changing to working directory and reading the file using pandas read_csv. 
The file is ; delimited and contains a header

In [2]:
os.chdir('E:\\Datasets')
os.listdir()

['airbnb-listings_netherlands.csv',
 'Churn_Modelling.csv',
 'creditcard.csv',
 'dataexam.txt',
 'F1',
 'housing',
 'Kaggle',
 'ml-100k',
 'model.pkl',
 'nlp',
 'PythonDA',
 'Sensor',
 'WideWorldImporters-Full.bak']

In [3]:
path = os.path.join(os.listdir()[0])
airbnb = pd.read_csv(path,delimiter=';',header=0) 
airbnb.head()


ID                            Listing Url       Scrape ID  \
0   9752227   https://www.airbnb.com/rooms/9752227  20170402075052   
1    711269    https://www.airbnb.com/rooms/711269  20170402075052   
2   7433172   https://www.airbnb.com/rooms/7433172  20170402075052   
3  16577285  https://www.airbnb.com/rooms/16577285  20170402075052   
4  16356202  https://www.airbnb.com/rooms/16356202  20170402075052   

  Last Scraped                                               Name  \
0   2017-04-02                   Amsterdam Suites, Art Deco suite   
1   2017-04-02                     Apt for Professionals/Families   
2   2017-04-02                  1 Br Apartment in Old South A'dam   
3   2017-04-02  AMSTERDAM OUD-ZUID/Vondelpark 25min. from Schi...   
4   2017-04-02  Warm and quiet single room in beautiful apartm...   

                                             Summary  \
0  A Bed & Breakfast Suite providing guests with ...   
1  Luxurious ,very light and spacious two-bedroom...   
2  A warm and spacious  1 bedroom apartment in Ol...   
3  Welkome to this spacious apartment located in ...   
4  Cosy bedroom in an apartment at the top of the...   

                                               Space  \
0  The suite offers state of the art entertainmen...   
1  Luxurious ,very light and spacious two-bedroom...   
2  The apartment is about 75 sqm and is located o...   
3  Apartment is clean and spacious with one bedro...   
4  Your bedroom is small but quiet and with good ...   

                                         Description Experiences Offered  \
0  A Bed & Breakfast Suite providing guests with ...                none   
1  Luxurious ,very light and spacious two-bedroom...                none   
2  A warm and spacious  1 bedroom apartment in Ol...                none   
3  Welkome to this spacious apartment located in ...                none   
4  Cosy bedroom in an apartment at the top of the...                none   

                               Neighborhood Overview  ...  \
0  Amsterdam Suites is situated in the heart of t...  ...   
1  The neighborhood is characterized by a spaciou...  ...   
2  The neighbourhood is wonderful and has tons of...  ...   
3  Bike rental is available on the corner of the ...  ...   
4  Schinkelbuurt is a beautiful zone in the west ...  ...   

  Review Scores Communication Review Scores Location Review Scores Value  \
0                        10.0                    8.0                 9.0   
1                        10.0                    9.0                 9.0   
2                         NaN                    NaN                 NaN   
3                         9.0                    9.0                 9.0   
4                        10.0                    9.0                10.0   

  License Jurisdiction Names Cancellation Policy  \
0     NaN          Amsterdam            moderate   
1     NaN          Amsterdam              strict   
2     NaN          Amsterdam              strict   
3     NaN          Amsterdam            moderate   
4     NaN          Amsterdam            flexible   

  Calculated host listings count Reviews per Month  \
0                              2              0.13   
1                              1              0.41   
2                              1               NaN   
3                              1              2.73   
4                              1              2.31   

                   Geolocation  \
0  52.3556516257,4.87922738637   
1  52.3452962399,4.86919559653   
2  52.3429194511,4.85961518996   
3    52.353563977,4.8549571139   
4  52.3523753503,4.85317422612   

                                            Features  
0  Host Has Profile Pic,Host Identity Verified,Is...  
1  Host Has Profile Pic,Host Identity Verified,Is...  
2             Host Has Profile Pic,Is Location Exact  
3                               Host Has Profile Pic  
4  Host Has Profile Pic,Host Identity Verified,Is...  

[5 rows x 89 columns]

In [4]:
airbnb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15181 entries, 0 to 15180
Data columns (total 89 columns):
ID                                15181 non-null int64
Listing Url                       15181 non-null object
Scrape ID                         15181 non-null int64
Last Scraped                      15181 non-null object
Name                              15181 non-null object
Summary                           14748 non-null object
Space                             11289 non-null object
Description                       15181 non-null object
Experiences Offered               15181 non-null object
Neighborhood Overview             10022 non-null object
Notes                             6506 non-null object
Transit                           10210 non-null object
Access                            9409 non-null object
Interaction                       9115 non-null object
House Rules                       10572 non-null object
Thumbnail Url                     11336 non-null object


##### Inspecting the dataset
The file contains 89 columns with 15,180 rows. 

In [5]:
airbnb.columns = airbnb.columns.str.replace(' ','_',) # replace whitespace with _ in the column header

In [6]:
airbnb[airbnb.duplicated() == True] == 0 # checking if any rows are duplicated

Empty DataFrame
Columns: [ID, Listing_Url, Scrape_ID, Last_Scraped, Name, Summary, Space, Description, Experiences_Offered, Neighborhood_Overview, Notes, Transit, Access, Interaction, House_Rules, Thumbnail_Url, Medium_Url, Picture_Url, XL_Picture_Url, Host_ID, Host_URL, Host_Name, Host_Since, Host_Location, Host_About, Host_Response_Time, Host_Response_Rate, Host_Acceptance_Rate, Host_Thumbnail_Url, Host_Picture_Url, Host_Neighbourhood, Host_Listings_Count, Host_Total_Listings_Count, Host_Verifications, Street, Neighbourhood, Neighbourhood_Cleansed, Neighbourhood_Group_Cleansed, City, State, Zipcode, Market, Smart_Location, Country_Code, Country, Latitude, Longitude, Property_Type, Room_Type, Accommodates, Bathrooms, Bedrooms, Beds, Bed_Type, Amenities, Square_Feet, Price, Weekly_Price, Monthly_Price, Security_Deposit, Cleaning_Fee, Guests_Included, Extra_People, Minimum_Nights, Maximum_Nights, Calendar_Updated, Has_Availability, Availability_30, Availability_60, Availability_90, Availability_365, Calendar_last_Scraped, Number_of_Reviews, First_Review, Last_Review, Review_Scores_Rating, Review_Scores_Accuracy, Review_Scores_Cleanliness, Review_Scores_Checkin, Review_Scores_Communication, Review_Scores_Location, Review_Scores_Value, License, Jurisdiction_Names, Cancellation_Policy, Calculated_host_listings_count, Reviews_per_Month, Geolocation, Features]
Index: []

[0 rows x 89 columns]

In [7]:
missing_val = airbnb.isnull().sum() /len(airbnb)
missing_val_gt_50 = airbnb.columns[missing_val > 0.5]

In [8]:
missing_val_gt_50

Index(['Notes', 'Host_Acceptance_Rate', 'Neighbourhood_Group_Cleansed',
       'Square_Feet', 'Weekly_Price', 'Monthly_Price', 'Has_Availability',
       'License'],
      dtype='object')

In [9]:
airbnb.drop(missing_val_gt_50,axis=1,inplace=True)

In [10]:
missing_val = airbnb.isnull().sum() /len(airbnb)
missing_val_gt_30 = airbnb.columns[missing_val > 0.3]
missing_val_gt_30

Index(['Neighborhood_Overview', 'Transit', 'Access', 'Interaction',
       'House_Rules', 'Host_About', 'Host_Neighbourhood', 'Neighbourhood',
       'Security_Deposit'],
      dtype='object')



Getting the columns which are of **numeric data type** and split them into a different column.

In [11]:
num_airbnb = airbnb.select_dtypes(include=np.number)
num_airbnb.columns

Index(['ID', 'Scrape_ID', 'Host_ID', 'Host_Response_Rate',
       'Host_Listings_Count', 'Host_Total_Listings_Count', 'Latitude',
       'Longitude', 'Accommodates', 'Bathrooms', 'Bedrooms', 'Beds', 'Price',
       'Security_Deposit', 'Cleaning_Fee', 'Guests_Included', 'Extra_People',
       'Minimum_Nights', 'Maximum_Nights', 'Availability_30',
       'Availability_60', 'Availability_90', 'Availability_365',
       'Number_of_Reviews', 'Review_Scores_Rating', 'Review_Scores_Accuracy',
       'Review_Scores_Cleanliness', 'Review_Scores_Checkin',
       'Review_Scores_Communication', 'Review_Scores_Location',
       'Review_Scores_Value', 'Calculated_host_listings_count',
       'Reviews_per_Month'],
      dtype='object')

Getting the columns which are of **categorical data type** and split them into a different column.

In [12]:
cat_airbnb = airbnb.select_dtypes(include=np.object)
cat_airbnb.columns

Index(['Listing_Url', 'Last_Scraped', 'Name', 'Summary', 'Space',
       'Description', 'Experiences_Offered', 'Neighborhood_Overview',
       'Transit', 'Access', 'Interaction', 'House_Rules', 'Thumbnail_Url',
       'Medium_Url', 'Picture_Url', 'XL_Picture_Url', 'Host_URL', 'Host_Name',
       'Host_Since', 'Host_Location', 'Host_About', 'Host_Response_Time',
       'Host_Thumbnail_Url', 'Host_Picture_Url', 'Host_Neighbourhood',
       'Host_Verifications', 'Street', 'Neighbourhood',
       'Neighbourhood_Cleansed', 'City', 'State', 'Zipcode', 'Market',
       'Smart_Location', 'Country_Code', 'Country', 'Property_Type',
       'Room_Type', 'Bed_Type', 'Amenities', 'Calendar_Updated',
       'Calendar_last_Scraped', 'First_Review', 'Last_Review',
       'Jurisdiction_Names', 'Cancellation_Policy', 'Geolocation', 'Features'],
      dtype='object')

In [13]:
num_airbnb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15181 entries, 0 to 15180
Data columns (total 33 columns):
ID                                15181 non-null int64
Scrape_ID                         15181 non-null int64
Host_ID                           15181 non-null int64
Host_Response_Rate                11535 non-null float64
Host_Listings_Count               15178 non-null float64
Host_Total_Listings_Count         15178 non-null float64
Latitude                          15181 non-null float64
Longitude                         15181 non-null float64
Accommodates                      15181 non-null int64
Bathrooms                         15163 non-null float64
Bedrooms                          15169 non-null float64
Beds                              15164 non-null float64
Price                             15173 non-null float64
Security_Deposit                  7752 non-null float64
Cleaning_Fee                      11468 non-null float64
Guests_Included                   15181 non-n

Lets take the mean for the numerical columns which are having Na values

In [14]:
num_airbnb = num_airbnb.fillna(num_airbnb.mean())

Correlation matrix for all the numerical values

In [15]:
correlation = num_airbnb.corr().round(2) 

Accommodates,Bedroom,Beds have higher correlation. When house has more bedroom,beds and more people to accomodate higher the price. 

In [16]:
num_airbnb.corr().round(2)['Price'].sort_values(ascending=False)

Price                             1.00
Accommodates                      0.57
Bedrooms                          0.54
Beds                              0.51
Cleaning_Fee                      0.44
Guests_Included                   0.37
Bathrooms                         0.33
Security_Deposit                  0.21
Availability_30                   0.18
Availability_60                   0.14
Review_Scores_Location            0.14
Extra_People                      0.13
Availability_365                  0.12
Availability_90                   0.11
Calculated_host_listings_count    0.11
Host_Total_Listings_Count         0.10
Host_Listings_Count               0.10
Review_Scores_Rating              0.06
Review_Scores_Cleanliness         0.04
Latitude                          0.03
Longitude                         0.03
Review_Scores_Accuracy            0.02
Review_Scores_Checkin             0.02
Review_Scores_Communication       0.02
Host_Response_Rate                0.01
Review_Scores_Value      

In [17]:
cat_airbnb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15181 entries, 0 to 15180
Data columns (total 48 columns):
Listing_Url               15181 non-null object
Last_Scraped              15181 non-null object
Name                      15181 non-null object
Summary                   14748 non-null object
Space                     11289 non-null object
Description               15181 non-null object
Experiences_Offered       15181 non-null object
Neighborhood_Overview     10022 non-null object
Transit                   10210 non-null object
Access                    9409 non-null object
Interaction               9115 non-null object
House_Rules               10572 non-null object
Thumbnail_Url             11336 non-null object
Medium_Url                11336 non-null object
Picture_Url               15167 non-null object
XL_Picture_Url            11336 non-null object
Host_URL                  15181 non-null object
Host_Name                 15178 non-null object
Host_Since                151

In [22]:
airbnb.City.value_counts()

Amsterdam                       14869
Amsterdam-Zuidoost                198
Amsterdam Zuid-Oost                54
De Pijp                             6
Jordaan                             5
Watergraafsmeer                     5
Amstelveen                          5
Amsterdam Zuidoost                  4
Ámsterdam                           4
Amsterdam (Zunderdorp)              3
Badhoevedorp                        3
Diemen                              2
Zeeburg                             2
Lijnden                             1
Amsterdão                           1
Amsterdam, Noord-Holland, NL        1
Haarlem                             1
Jordaan, Amsterdam                  1
Oud Ade                             1
1054 Amsterdam                      1
AMSTERDAM                           1
Am                                  1
North Holland                       1
Buitenveldert                       1
Centrem                             1
Nederland                           1
Duivendrecht

In [130]:
airbnb.City.apply(lambda x: str(x).lower().replace(regex=r'mster',value='amsterdam'))

TypeError: replace() takes no keyword arguments

In [105]:
import string 
string.punctuation
airbnb.State.replace()

True


In [146]:
airbnb.State = airbnb.State.str.lower().replace(
    ['noord holland',
     'north holland',
     'north-holland',
     'noord holland',
     'nh','amsterdam',
     'netherlands'
     r'amster',
     r'holland$'
    ],
    'holland',regex=True)

In [147]:
airbnb.State.value_counts()

holland                 15092
am                          2
van heemskeckstraat         1
kx                          1
nr                          1
                        ...  
dp                          1
jt                          1
emmy andriessestraat        1
jan tooropstraat            1
hc                          1
Name: State, Length: 63, dtype: int64

In [179]:
airbnb.State.unique()

array(['holland', 'ng', 'bw', 'blancefloorstraat', 'ch', nan, 'am', 'gj',
       'sx', 'ke', 'slotermeerlaan', 'kx', '244', 'holanda septentrional',
       'ka', 'ah', 'van heemskeckstraat', 'modern', 'transvaalstraat',
       'vl', 'nr', 'hunzestraat', 'tl', 'gn', 'ce', 'rj',
       'valckenierstraat 1', 'zeedijk', 'dp', 'kw', 'jan tooropstraat',
       'ln', 'nc', 'amstel', 'jt', 'nb', 'boord holland', 'ts',
       'noordholland', 'mj', 'pieter calandlaan', 'vx',
       'fritz dietrich kahlenbergstraat', 'south holland', 'insulindeweg',
       'zuiderzeeweg', 'wx', 'xh', 'nah', 'jordaan holland',
       '2 de egelantiersdwarsstraat 13 iii', 'hc', 'gd', 'ga', 'nw',
       'emmy andriessestraat', 'nassaukade', 'nederland', '38c',
       'nieuwe hoogstraat', 'daniel stalpertstraat', 'borgerstraat',
       'ijburglaan', 'nm'], dtype=object)

In [218]:
airbnb.Host_Location.replace(r'(^Amsterdam.)','Amsterdam',regex=True)

0                                     Netherlands
1        Amsterdam North Holland, The Netherlands
2                       Shanghai, Shanghai, China
3            Amsterdam North Holland, Netherlands
4            Amsterdam North Holland, Netherlands
                           ...                   
15176    Amsterdam Noord-Holland, The Netherlands
15177        Amsterdam North Holland, Netherlands
15178        Amsterdam North Holland, Netherlands
15179    Amsterdam North Holland, The Netherlands
15180        Amsterdam North Holland, Netherlands
Name: Host_Location, Length: 15181, dtype: object

Grouping the list of features to impute and clean values.

In [200]:
airbnb.Host_Location.value_counts()

Amsterdam, North Holland, Netherlands        6826
Amsterdam, North Holland, The Netherlands    4458
Amsterdam, Noord-Holland, The Netherlands    1666
NL                                           1155
Netherlands                                   115
                                             ... 
Cesena, Emilia-Romagna, Italy                   1
San Francisco, California, United States        1
Empe, Gelderland, Netherlands                   1
Amsterdan                                       1
Hoorn, North Holland, The Netherlands           1
Name: Host_Location, Length: 409, dtype: int64

In [156]:
categorical_list = ['Property_Type',
'Room_Type',
'Bed_Type',
'Cancellation_Policy',
'Host_Response_Time',
'Neighbourhood_Cleansed',
'City',
'Host_Location',
'Host_Neighbourhood',
'Neighbourhood',
'State'
]

replace_na_with_zero = ['Extra_People',
'Minimum_Nights',
'Maximum_Nights',
'Review_Scores_Cleanliness',
'Review_Scores_Checkin',
'Review_Scores_Communication',
'Review_Scores_Location',
'Review_Scores_Value',
'Bathrooms',
'Bedrooms',
'Beds',
'Price',
'Cleaning_Fee',
'Review_Scores_Rating',
'Review_Scores_Accuracy',
'Reviews_per_Month',
'Security_Deposit',
'Number_of_Reviews',
'Calculated_host_listings_count',
'Accommodates',
'Guests_Included',
'Availability_30',
'Availability_60',
'Availability_90',
'Availability_365',
'Latitude',
'Longitude'
]

replace_val_with_0_or_1 = ['Summary',
'Space',
'Neighborhood_Overview',
'Interaction',
'Thumbnail_Url',
'Medium_Url',
'Picture_Url',
'XL_Picture_Url',
'Host_URL',
'Host_Name',
'Host_About',
'Host_Thumbnail_Url',
'Host_Picture_Url',
'Transit',
'Access',
'House_Rules'
]


replace_with_count_values = ['Host_Verifications',
'Amenities',
'Features'
]

replace_na_with_1 = ['Host_Listings_Count',
'Host_Total_Listings_Count'
]

replace_date_values = ['Host_Since',
'First_Review',
'Last_Review'
]

In [161]:
categorical_list = airbnb[categorical_list].replace(np.nan,'Not Available')
categorical_list = pd.get_dummies(categorical_list).astype('int8')
categorical_list

ValueError: Must pass DataFrame with boolean values only

In [165]:
replace_na_with_zero = airbnb[replace_na_with_zero].replace(np.nan,0)
replace_na_with_zero.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15181 entries, 0 to 15180
Data columns (total 27 columns):
Extra_People                      15181 non-null int64
Minimum_Nights                    15181 non-null int64
Maximum_Nights                    15181 non-null int64
Review_Scores_Cleanliness         15181 non-null float64
Review_Scores_Checkin             15181 non-null float64
Review_Scores_Communication       15181 non-null float64
Review_Scores_Location            15181 non-null float64
Review_Scores_Value               15181 non-null float64
Bathrooms                         15181 non-null float64
Bedrooms                          15181 non-null float64
Beds                              15181 non-null float64
Price                             15181 non-null float64
Cleaning_Fee                      15181 non-null float64
Review_Scores_Rating              15181 non-null float64
Review_Scores_Accuracy            15181 non-null float64
Reviews_per_Month                 15181 no

In [166]:
replace_val_with_0_or_1 = airbnb[replace_val_with_0_or_1].apply(lambda col: pd.notna(col).astype('int8'))
replace_val_with_0_or_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15181 entries, 0 to 15180
Data columns (total 16 columns):
Summary                  15181 non-null int8
Space                    15181 non-null int8
Neighborhood_Overview    15181 non-null int8
Interaction              15181 non-null int8
Thumbnail_Url            15181 non-null int8
Medium_Url               15181 non-null int8
Picture_Url              15181 non-null int8
XL_Picture_Url           15181 non-null int8
Host_URL                 15181 non-null int8
Host_Name                15181 non-null int8
Host_About               15181 non-null int8
Host_Thumbnail_Url       15181 non-null int8
Host_Picture_Url         15181 non-null int8
Transit                  15181 non-null int8
Access                   15181 non-null int8
House_Rules              15181 non-null int8
dtypes: int8(16)
memory usage: 237.3 KB


In [167]:
replace_with_count_values = airbnb[replace_with_count_values].apply(lambda col: col.str.split(',').str.len().fillna(0).astype('int8'))
replace_with_count_values.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15181 entries, 0 to 15180
Data columns (total 3 columns):
Host_Verifications    15181 non-null int8
Amenities             15181 non-null int8
Features              15181 non-null int8
dtypes: int8(3)
memory usage: 44.6 KB


In [168]:
replace_na_with_1 = airbnb[replace_na_with_1].replace(np.nan,1)
replace_na_with_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15181 entries, 0 to 15180
Data columns (total 2 columns):
Host_Listings_Count          15181 non-null float64
Host_Total_Listings_Count    15181 non-null float64
dtypes: float64(2)
memory usage: 237.3 KB


In [169]:
replace_date_values = airbnb[replace_date_values].fillna(method='ffill')
replace_date_values = replace_date_values.apply(lambda col: pd.factorize(col,sort=True)[0].astype('int'))
replace_date_values.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15181 entries, 0 to 15180
Data columns (total 3 columns):
Host_Since      15181 non-null int32
First_Review    15181 non-null int32
Last_Review     15181 non-null int32
dtypes: int32(3)
memory usage: 178.0 KB


In [170]:
updated = pd.concat([categorical_list,
           replace_na_with_zero,
           replace_val_with_0_or_1,
           replace_with_count_values,
           replace_na_with_1,
           replace_date_values
          ],axis=1)

In [171]:
updated.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15181 entries, 0 to 15180
Columns: 729 entries, Property_Type_Apartment to Last_Review
dtypes: float64(18), int32(3), int64(11), int8(697)
memory usage: 13.6 MB


In [177]:
correlation = updated.corr()

In [178]:
correlation['Price'][correlation['Price'] > 0.1].sort_values(ascending=False)

Price                                  1.000000
Accommodates                           0.562822
Bedrooms                               0.537698
Beds                                   0.511277
Guests_Included                        0.364912
Cleaning_Fee                           0.364290
Bathrooms                              0.324807
Room_Type_Entire home/apt              0.316342
Security_Deposit                       0.226117
Neighbourhood_Cleansed_Centrum-West    0.190903
Amenities                              0.187655
Availability_30                        0.174742
Cancellation_Policy_strict             0.146660
Availability_60                        0.139785
Extra_People                           0.129945
Neighbourhood_Grachtengordel           0.124031
Availability_365                       0.119076
Host_Neighbourhood_Grachtengordel      0.116178
Availability_90                        0.111759
Calculated_host_listings_count         0.108775
Property_Type_House                    0

In [ ]:
from sklearn.feature_selection import SelectKBest,f_regression
sel = SelectKBest(f_regression,k=89)

In [ ]:
val = correlation[correlation > 0.5] 

In [175]:
updated['updated_cat'] = pd.cut(updated['Accommodates'],
                    bins=[0,1,2,3,4,8,12,np.inf],
                    labels=[1,2,3,4,5,6,7])
# binning the category in order to split train and test data based on the accommodates to get a stratified split

In [ ]:
from sklearn.model_selection import StratifiedShuffleSplit
strata = StratifiedShuffleSplit(n_splits=1,test_size=0.2,random_state=42)
for train_size,test_size in strata.split(updated,updated['updated_cat']):
    train_set = updated.iloc[train_size]
    test_set = updated.iloc[test_size]

In [ ]:
train_set['updated_cat'].value_counts() / len(updated)

In [ ]:
data = train_set.drop(['Price','updated_cat'],axis=1)
price = train_set['Price'].copy()

In [ ]:
from sklearn.preprocessing import StandardScaler
stdsclr = StandardScaler()
scaled = stdsclr.fit_transform(data)

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
rforest_reg = RandomForestRegressor(n_jobs=-1,random_state=42)

In [ ]:
rforest_reg.fit(scaled,price)

In [ ]:
rpredictor = rforest_reg.predict(scaled)
mse = mean_squared_error(price,rpredictor)
rmse = np.sqrt(mse)
rmse

In [ ]:
from sklearn.model_selection import cross_val_score
rand_score = cross_val_score(rforest_reg,scaled,price,
                            scoring='neg_mean_squared_error',cv=10,n_jobs=-1)
forest_rmse = np.sqrt(-rand_score)

In [ ]:
np.mean(forest_rmse),np.std(forest_rmse)

In [ ]:
from scipy import stats
confidence = 0.95
squared_errors = (rpredictor - price)**2

In [ ]:
np.sqrt(stats.t.interval(confidence,len(squared_errors) -1,loc=squared_errors.mean(),
                 scale=stats.sem(squared_errors)))